### PR TITLE
Use ModelBuilder for incrementality model

### PR DESCRIPTION
**Pull request by wd60622**
Use the ModelBuilder for #803

Also uses the Prior class to generalize the logic a bit. The previous priors can be created with:

```python
def set_priors_from_existing_sales(model, existing_sales):
    model.model_config["intercept"].parameters["mu"] = existing_sales.mean().values
    model.model_config["intercept"].parameters["sigma"] = existing_sales.std().values
    model.model_config["likelihood"]["sigma"].parameters["sigma"] = (
        existing_sales.std().mean()
    )
```

Using the ModelBuilder allows user to save and load the model.

```python
test_file = "test-mvits.nc"
model.save(test_file)

loaded = MVITS.load(test_file)
```

CC: @juanitorduz @drbenvincent

----
📚 Documentation preview 📚: https://pymc-marketing--1101.org.readthedocs.build/en/1101/

### CODE REVIEW

File: pymc_marketing/model_builder.py [hunk 1]
```diff
@@ -580,11 +580,17 @@ def fit(
         Initializing NUTS using jitter+adapt_diag...
 
         """
+        if isinstance(y, pd.Series) and not X.index.equals(y.index):
```
Comment (juanitorduz) [inline pymc_marketing/model_builder.py:583]: can we add a test? :)

File: pymc_marketing/product_incrementality/mv_its.py [hunk 2]
```diff
@@ -13,176 +13,217 @@
...
+        """Default model configuration."""
+        return {
+            "intercept": Prior("Normal", dims="background_product"),
+            "likelihood": Prior(
+                "TruncatedNormal",
+                lower=0,
+                sigma=Prior("HalfNormal", dims="background_product"),
+                dims=("time", "background_product"),
+            ),
+            "alpha_background": 0.5,
```
Comment (juanitorduz) [inline pymc_marketing/product_incrementality/mv_its.py:80]: what is the meaning of this `alpha_background` :D ?

File: pymc_marketing/product_incrementality/mv_its.py [hunk 3]
```diff
@@ -13,176 +13,217 @@
...
-                    var_names=["mu", "y"],
-                    random_seed=self.rng,
+                    var_names=["mu", self.output_var],
+                    random_seed=random_seed,
                     predictions=True,
                 )
             )
 
-    def causal_impact(self, variable="mu"):
+    def sample(self, X, y, random_seed: int | None = None):
```
Comment (juanitorduz) [inline pymc_marketing/product_incrementality/mv_its.py:213]: can we add a self-like return type?

**Review by juanitorduz — COMMENTED** [review]
Thanks @wd60622 ! I think this is a great addition! It makes the code more cohesive to what we already have. I have added a couple of small comments and asked for a small test ;)

File: pymc_marketing/product_incrementality/mv_its.py (hunk 2 above)
Comment (wd60622) [reply pymc_marketing/product_incrementality/mv_its.py:80]: @drbenvincent

Comment (wd60622) [reply pymc_marketing/product_incrementality/mv_its.py]: I have moved this entirely to the `model_config` which will have a default like before but it also exposes it to the user to create (with some limitations)

File: pymc_marketing/product_incrementality/mv_its.py [hunk 4]
```diff
@@ -13,176 +13,255 @@
...
+
+        return result
+
+    def _generate_and_preprocess_model_data(
+        self,
+        X: pd.DataFrame | pd.Series,
+        y: np.ndarray,
+    ) -> None:
+        if isinstance(X, pd.Series):
+            raise ValueError("X must be a DataFrame, not a Series")  # pragma: no cover
```
Comment (wd60622) [inline pymc_marketing/product_incrementality/mv_its.py:132]: Message doesnt make sense

**Review by wd60622 — COMMENTED** [review]
Thoughts on this API? Does using "y" for treatment sales make sense or should that be part of X as well? 

Would this treatment sales ever be a DataFrame as well? Is that an extension we foresee?

**Review by drbenvincent — CHANGES_REQUESTED** [review]
Thanks for the updates.

I can get the tests in `test_mv_its.py` to pass.

However, the notebooks (`mv_its_saturated.ipynb` and `mv_its_unsaturated.ipynb`) seem to not work:
* `treatment_time` seems to be absent as a kwarg from `MVITS.__init__`
* `MVITS.plot_data` is called in the notebook like a static method, but that isn't working / won't work

Would you be able to update these docs notebooks with the new API?

**Comment by wd60622** [comment]
> Thanks for the updates 🥳
> 
> 
> 
> I can get the tests in `test_mv_its.py` to pass.
> 
> 
> 
> However, the notebooks (`mv_its_saturated.ipynb` and `mv_its_unsaturated.ipynb`) seem to not work:
> 
> * `treatment_time` seems to be absent as a kwarg from `MVITS.__init__`
> 
> * `MVITS.plot_data` is called in the notebook like a static method, but that isn't working / won't work
> 
> 
> 
> Would you be able to update these docs notebooks with the new API?

Yes, I can make those changes. The full CI suite isnt running as far as I can tell. Shall I merge into your previous branch and continue there? @drbenvincent

**Comment by drbenvincent** [comment]
> Yes, I can make those changes. The full CI suite isnt running as far as I can tell. Shall I merge into your previous branch and continue there? @drbenvincent

Yes, makes sense. It was good to branch off of the branch, but I think it's clear that we do want this inheritance from `ModelBuilder`. As you say, more refinement can be done in the original `incrementality` branch.

**Pull request merged**